cpp/src/arrow/filesystem/test_util.cc

```diff
@@ -111,6 +113,12 @@ void SortInfos(std::vector<FileInfo>* infos) {
   std::sort(infos->begin(), infos->end(), FileInfo::ByPath{});
 }
 
+void AssertFileInfoGenerator(FileInfoGenerator gen, FileInfoVector* out_infos) {
```

Comment:
Naming nit: Perhaps don't name this `Assert` since it isn't actually asserting the contents but just collecting them?  I see ASSERT_FINISHES_OK_AND_ASSIGN as an analogue of ASSERT_OK_AND_ASSIGN which is more of a utility assertion and less of a test expectation.

Reply:
Perhaps `CollectFileInfoGenerator`?

Reply:
That works.  Also, minor nit, I suppose you could `std::move(nested_infos)` into the call to `FlattenVectors`.